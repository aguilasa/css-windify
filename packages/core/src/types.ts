/**
 * Type definitions for Tailwindify
 */

export interface TailwindifyOptions {
  prefix?: string;
  strict?: boolean;
  approximate?: boolean;
}

export interface CssDeclaration {
  prop: string;
  value: string;
  variants?: string[];
}

export interface CssRule {
  selector: string;
  declarations: CssDeclaration[];
}

export type PropertyCategory =
  | 'spacing'
  | 'color'
  | 'typography'
  | 'layout'
  | 'border'
  | 'background'
  | 'effects'
  | 'other';

export type WarningCategory = 'arbitrary-value' | 'no-handler' | 'approximate' | 'other';

export interface CategoryStats {
  matched: number;
  total: number;
  percentage: number;
}

export interface TransformResult {
  classes: string[];
  warnings: string[];
  coverage: {
    matched: number;
    total: number;
    percentage: number;
    nonArbitrary: number;
    categories?: Record<PropertyCategory, CategoryStats>;
    warningsByCategory?: Record<WarningCategory, number>;
  };
}

export interface MatchCtx {
  theme: any;
  opts: {
    strict: boolean;
    approximate: boolean;
  };
}

export type RuleHandler = (
  value: string,
  ctx: MatchCtx
) => string[] | { classes: string[]; warnings: string[] } | null;
