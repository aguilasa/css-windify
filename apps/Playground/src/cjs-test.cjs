// CommonJS test
const { transformDeclarations } = require('tailwindify-core');

// Test data
const declarations = [
  { prop: 'margin', value: '1rem' },
  { prop: 'padding', value: '2rem' }
];

const ctx = {
  theme: {},
  opts: {
    strict: false,
    approximate: true
  }
};

// Run the transformation
const result = transformDeclarations(declarations, ctx);

// Output results
console.log(`CJS Test - Classes: ${result.classes.join(', ')}`);
console.log(`CJS Test - Warnings: ${result.warnings.join(', ')}`);
console.log(`CJS Test - Coverage: ${result.coverage.matched}/${result.coverage.total}`);
