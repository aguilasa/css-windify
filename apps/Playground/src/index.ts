import { Tailwindify } from 'tailwindify-core';

// Simple demo of the Tailwindify core library
function runDemo() {
  console.log('Initializing Tailwindify Playground...');

  const tailwindify = new Tailwindify({ prefix: 'tw-' });

  const input = 'button';
  const output = tailwindify.process(input);

  console.log(`Input: "${input}"`);
  console.log(`Output: "${output}"`);
  console.log('Playground ready');
}

// Run the demo
runDemo();
