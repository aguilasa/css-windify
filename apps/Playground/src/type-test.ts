import { transformDeclarations, TransformResult, CssDeclaration, MatchCtx } from 'css-windify-core';

// Teste de tipos
const declarations: CssDeclaration[] = [
  { prop: 'margin', value: '1rem' },
  { prop: 'padding', value: '2rem' },
];

const ctx: MatchCtx = {
  theme: {},
  opts: {
    strict: false,
    approximate: true,
  },
};

const result: TransformResult = transformDeclarations(declarations, ctx);

// Verificar se os tipos est√£o corretos
console.log(`Classes: ${result.classes.join(', ')}`);
console.log(`Warnings: ${result.warnings.join(', ')}`);
console.log(`Coverage: ${result.coverage.matched}/${result.coverage.total}`);
