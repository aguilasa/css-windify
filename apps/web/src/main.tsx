/**
 * Main entry point for the Tailwindify Web application
 * This is a placeholder file that will be implemented in the future
 */

console.log('Tailwindify Web - Coming soon!');
