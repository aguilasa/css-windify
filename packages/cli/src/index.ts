#!/usr/bin/env node

/**
 * Tailwindify CLI
 * A command-line tool for converting CSS to Tailwind CSS
 */

console.log('Tailwindify CLI - Coming soon!');
