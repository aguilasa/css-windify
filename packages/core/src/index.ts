/**
 * Tailwindify core functionality
 */
import { transformRule, transformDeclarations } from './core/rulesEngine';
import { TailwindifyOptions, CssRule, CssDeclaration, MatchCtx, TransformResult } from './types';

// Re-export types
export * from './types';

class Tailwindify {
  private options: TailwindifyOptions;

  constructor(options: TailwindifyOptions = {}) {
    this.options = {
      prefix: '',
      strict: true,
      approximate: false,
      ...options,
    };
  }

  /**
   * Convert a CSS rule to Tailwind classes
   */
  processRule(rule: CssRule): TransformResult {
    const ctx: MatchCtx = {
      theme: {}, // Theme would be loaded from Tailwind config
      opts: {
        strict: !!this.options.strict,
        approximate: !!this.options.approximate
      }
    };
    
    const result = transformRule(rule, ctx);
    
    // Apply prefix if specified
    if (this.options.prefix) {
      result.classes = result.classes.map(cls => `${this.options.prefix}${cls}`);
    }
    
    return result;
  }

  /**
   * Process a list of CSS declarations
   */
  processDeclarations(declarations: CssDeclaration[]): TransformResult {
    const ctx: MatchCtx = {
      theme: {}, // Theme would be loaded from Tailwind config
      opts: {
        strict: !!this.options.strict,
        approximate: !!this.options.approximate
      }
    };
    
    const result = transformDeclarations(declarations, ctx);
    
    // Apply prefix if specified
    if (this.options.prefix) {
      result.classes = result.classes.map(cls => `${this.options.prefix}${cls}`);
    }
    
    return result;
  }
}

// Export types and functions
export { toTailwind } from './core/rulesEngine';
export { transformRule, transformDeclarations };
export {
  normalizeValue,
  isPx,
  isRem,
  isEm,
  isPct,
  isNumber,
  toArbitrary,
  arbitraryProperty,
  parseBoxShorthand,
  parseColorNormalize
} from './core/normalizers';

export {
  loadTheme,
  defaultTheme,
  resolveSpacingToken,
  resolveColorToken,
  resolveFontSizeToken,
  resolveLineHeightToken
} from './core/themeLoader';

export {
  matchSpacing,
  matchTypography,
  matchDisplay,
  matchPosition,
  matchInset,
  matchInsetShorthand,
  displayMap,
  positionMap
} from './core/matchers';

// Export both named and default exports
export { Tailwindify };
export default Tailwindify;
