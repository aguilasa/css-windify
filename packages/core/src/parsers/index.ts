/**
 * Export all parsers
 */

export * from './inlineCss';
export * from './cssRules';
