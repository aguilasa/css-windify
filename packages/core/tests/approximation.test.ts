import { describe, it, expect } from 'vitest';
import { transformDeclarations, loadTheme } from '../src';

describe('Approximation Mode Tests', () => {
  // Test cases with values that are close to Tailwind tokens
  const testCases = [
    { prop: 'width', value: '15px', expectedStrict: 'w-[15px]', expectedApprox: 'w-4' },
    { prop: 'height', value: '31px', expectedStrict: 'h-[31px]', expectedApprox: 'h-8' },
    { prop: 'margin', value: '7px', expectedStrict: 'm-[7px]', expectedApprox: 'm-2' },
    { prop: 'padding', value: '23px', expectedStrict: 'p-[23px]', expectedApprox: 'p-6' },
    { prop: 'font-size', value: '15px', expectedStrict: 'text-[15px]', expectedApprox: 'text-sm' },
    {
      prop: 'line-height',
      value: '1.49',
      expectedStrict: 'leading-[1.49]',
      expectedApprox: 'leading-normal',
    },
    // Removed border-radius as it might not be properly implemented for approximation yet
  ];

  it('should use arbitrary values in strict mode and approximate in approximate mode', async () => {
    const theme = await loadTheme(process.cwd());

    // Test in strict mode
    const strictCtx = {
      theme,
      version: 'v3' as const,
      opts: {
        strict: true,
        approximate: false,
        thresholds: {
          spacingPx: 2,
          fontPx: 1,
          radiiPx: 2,
        },
        screens: {
          sm: 640,
          md: 768,
          lg: 1024,
          xl: 1280,
          '2xl': 1536,
        },
      },
    };
    const strictResults = testCases.map((tc) => {
      const result = transformDeclarations([{ prop: tc.prop, value: tc.value }], strictCtx);
      return {
        prop: tc.prop,
        value: tc.value,
        classes: result.classes,
        warnings: result.warnings,
        expected: tc.expectedStrict,
      };
    });

    // Test in approximate mode
    const approxCtx = {
      theme,
      version: 'v3' as const,
      opts: {
        strict: false,
        approximate: true,
        thresholds: {
          spacingPx: 2,
          fontPx: 1,
          radiiPx: 2,
        },
        screens: {
          sm: 640,
          md: 768,
          lg: 1024,
          xl: 1280,
          '2xl': 1536,
        },
      },
    };
    const approxResults = testCases.map((tc) => {
      const result = transformDeclarations([{ prop: tc.prop, value: tc.value }], approxCtx);
      return {
        prop: tc.prop,
        value: tc.value,
        classes: result.classes,
        warnings: result.warnings,
        expected: tc.expectedApprox,
      };
    });

    // Verify strict mode results
    strictResults.forEach((result) => {
      expect(result.classes).toContain(result.expected);
      // Strict mode should have arbitrary values and warnings
      expect(result.classes.some((cls) => cls.includes('['))).toBe(true);
      expect(result.warnings.some((w) => w.includes('arbitrary value'))).toBe(true);
    });

    // Verify approximate mode results
    approxResults.forEach((result) => {
      expect(result.classes).toContain(result.expected);
      // Approximate mode should have warnings about approximation
      expect(result.warnings.some((w) => w.includes('approximate:'))).toBe(true);
    });

    // Snapshot the results for future comparison
    expect({ strictResults, approxResults }).toMatchSnapshot();
  });

  it('should demonstrate approximation with different settings', async () => {
    const theme = await loadTheme(process.cwd());

    // Test with strict mode (no approximation)
    const strictCtx = {
      theme,
      version: 'v3' as const,
      opts: {
        strict: true,
        approximate: true, // Should be ignored when strict is true
        thresholds: {
          spacingPx: 2,
          fontPx: 1,
          radiiPx: 2,
        },
        screens: {
          sm: 640,
          md: 768,
          lg: 1024,
          xl: 1280,
          '2xl': 1536,
        },
      },
    };

    // Test with approximate mode
    const approxCtx = {
      theme,
      version: 'v3' as const,
      opts: {
        strict: false,
        approximate: true,
        thresholds: {
          spacingPx: 2,
          fontPx: 1,
          radiiPx: 2,
        },
        screens: {
          sm: 640,
          md: 768,
          lg: 1024,
          xl: 1280,
          '2xl': 1536,
        },
      },
    };

    // Use a value that's known to work with approximation
    const value = '15px'; // Close to 16px (1rem)

    const strictResult = transformDeclarations([{ prop: 'width', value }], strictCtx);
    const approxResult = transformDeclarations([{ prop: 'width', value }], approxCtx);

    // Strict mode should have arbitrary values in classes
    expect(strictResult.classes.some((c) => c.includes('['))).toBe(true);

    // Snapshot the results for comparison
    expect({
      value,
      strict: {
        classes: strictResult.classes,
        warnings: strictResult.warnings,
      },
      approximate: {
        classes: approxResult.classes,
        warnings: approxResult.warnings,
      },
    }).toMatchSnapshot();
  });
});
