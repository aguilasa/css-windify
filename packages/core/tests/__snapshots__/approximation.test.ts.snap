// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Approximation Mode Tests > should demonstrate approximation with different settings 1`] = `
{
  "approximate": {
    "classes": [
      "w-4",
    ],
    "warnings": [
      "approximate mapping: 15px → w-4 (1px difference)",
    ],
  },
  "strict": {
    "classes": [
      "w-[15px]",
    ],
    "warnings": [
      "Used arbitrary value for 'width: 15px'",
    ],
  },
  "value": "15px",
}
`;

exports[`Approximation Mode Tests > should use arbitrary values in strict mode and approximate in approximate mode 1`] = `
{
  "approxResults": [
    {
      "classes": [
        "w-4",
      ],
      "expected": "w-4",
      "prop": "width",
      "value": "15px",
      "warnings": [
        "approximate mapping: 15px → w-4 (1px difference)",
      ],
    },
    {
      "classes": [
        "h-8",
      ],
      "expected": "h-8",
      "prop": "height",
      "value": "31px",
      "warnings": [
        "approximate mapping: 31px → h-8 (1px difference)",
      ],
    },
    {
      "classes": [
        "m-2",
      ],
      "expected": "m-2",
      "prop": "margin",
      "value": "7px",
      "warnings": [
        "approximate mapping: 7px → m-2 (1px difference)",
      ],
    },
    {
      "classes": [
        "p-6",
      ],
      "expected": "p-6",
      "prop": "padding",
      "value": "23px",
      "warnings": [
        "approximate mapping: 23px → p-6 (1px difference)",
      ],
    },
    {
      "classes": [
        "text-sm",
      ],
      "expected": "text-sm",
      "prop": "font-size",
      "value": "15px",
      "warnings": [
        "approximate mapping: 15px → text-sm (1px difference)",
      ],
    },
    {
      "classes": [
        "leading-normal",
      ],
      "expected": "leading-normal",
      "prop": "line-height",
      "value": "1.49",
      "warnings": [
        "approximate mapping: 1.49 → leading-normal (0.010000000000000009px difference)",
      ],
    },
  ],
  "strictResults": [
    {
      "classes": [
        "w-[15px]",
      ],
      "expected": "w-[15px]",
      "prop": "width",
      "value": "15px",
      "warnings": [
        "Used arbitrary value for 'width: 15px'",
      ],
    },
    {
      "classes": [
        "h-[31px]",
      ],
      "expected": "h-[31px]",
      "prop": "height",
      "value": "31px",
      "warnings": [
        "Used arbitrary value for 'height: 31px'",
      ],
    },
    {
      "classes": [
        "m-[7px]",
      ],
      "expected": "m-[7px]",
      "prop": "margin",
      "value": "7px",
      "warnings": [
        "Used arbitrary value for 'margin: 7px'",
      ],
    },
    {
      "classes": [
        "p-[23px]",
      ],
      "expected": "p-[23px]",
      "prop": "padding",
      "value": "23px",
      "warnings": [
        "Used arbitrary value for 'padding: 23px'",
      ],
    },
    {
      "classes": [
        "text-[15px]",
      ],
      "expected": "text-[15px]",
      "prop": "font-size",
      "value": "15px",
      "warnings": [
        "Used arbitrary value for 'font-size: 15px'",
      ],
    },
    {
      "classes": [
        "leading-[1.49]",
      ],
      "expected": "leading-[1.49]",
      "prop": "line-height",
      "value": "1.49",
      "warnings": [
        "Used arbitrary value for 'line-height: 1.49'",
      ],
    },
  ],
}
`;
